src/Kerbalism/SteppedSim/FluxAnalysisFactory.cs

```diff
@@ -0,0 +1,223 @@
+			var bodySunOcclusionJob = new BodySunOcclusionJob
+			{
+				triplets = triplets,
+				bodies = bodies,
+				bodiesPerStep = numBodies,
+				starIndexes = starIndex.AsDeferredJobArray(),
+				occluded = bodyOccludedFromSun,
+			}.Schedule(numSteps * numBodies, 16, JobHandle.CombineDependencies(buildTripletsJob, gatherStarsJob));
+
+			vesselBodyOcclusionMap  = new NativeArray<bool>(fullSize, Allocator.TempJob);
+			var vesselBodyOcclusionJob = new BodyVesselOcclusionJob
+			{
+				triplets = triplets,
+				numBodiesPerStep = numBodies,
+				bodies = bodies,
+				vessels = vessels,
+				occlusionRelevance = occlusionRelevance,
+				occluded = vesselBodyOcclusionMap,
+			}.Schedule(fullSize, 16, fluxFactsJob);
+
+			var bodyFlux = new NativeArray<double>(fullSize, Allocator.TempJob);
+			var bodyFluxJob = new BodyEmissiveFlux
+			{
+				triplets = triplets,
+				bodies = bodies,
+				sunFluxAtBodies = sunFluxAtBodies,
+				flux = bodyFlux,
+			}.Schedule(fullSize, 256, sunFluxAtBodyJob);
+
+			var directIrradiance = new NativeArray<double>(fullSize, Allocator.TempJob);
+			var coreIrradiance = new NativeArray<double>(fullSize, Allocator.TempJob);
+			var emissiveIrradiance = new NativeArray<double>(fullSize, Allocator.TempJob);
+			var bodyDirectIrradiancesJob = new BodyDirectIrradiances
+			{
+				triplets = triplets,
+				bodies = bodies,
+				distances = vesselBodyDistance,
+				bodyEmissiveFlux = bodyFlux,
+				occluded = vesselBodyOcclusionMap,
+				solarIrradiance = directIrradiance,
+				bodyCoreIrradiance = coreIrradiance,
+				bodyEmissiveIrradiance = emissiveIrradiance,
+			}.Schedule(fullSize, 256, JobHandle.CombineDependencies(vesselBodyOcclusionJob, bodyFluxJob, fluxFactsJob));
+
+			var bodyHemisphericReradiatedIrradianceFactor = new NativeArray<double>(numSteps * numBodies, Allocator.TempJob);
+			var bodyHemisphericReradiatedIrradianceFactorJob = new BodyHemisphericReradiatedIrradianceFactor
+			{
+				bodies = bodies,
+				occluded = bodyOccludedFromSun,
+				sunFluxAtBody = sunFluxAtBodies,
+				factor = bodyHemisphericReradiatedIrradianceFactor
+			}.Schedule(numSteps * numBodies, 64, JobHandle.CombineDependencies(sunFluxAtBodyJob, bodySunOcclusionJob));
+
+			var albedoIrradiance = new NativeArray<double>(fullSize, Allocator.TempJob);
+			var albedoIrradianceAtVesselJob = new AlbedoIrradianceAtVesselJob
+			{
+				triplets = triplets,
+				numBodiesPerStep = numBodies,
+				bodies = bodies,
+				vessels = vessels,
+				starIndexes = starIndex.AsDeferredJobArray(),
+				directions = vesselBodyDirection,
+				distances = vesselBodyDistance,
+				vesselOccludedFromBody = vesselBodyOcclusionMap,
+				hemisphericReradiatedIrradianceFactor = bodyHemisphericReradiatedIrradianceFactor,
+				irradiance = albedoIrradiance,
+			}.Schedule(fullSize, 256, JobHandle.CombineDependencies(bodyHemisphericReradiatedIrradianceFactorJob, vesselBodyOcclusionJob));
+
```

Comment:
In some places, yes.  Here, I'm pretty confident.  But you need to see what each job is doing.
In this example, for the albedo computation, yes, it makes sense to use the total body flux at the body and pre-calculate the flux/luminosity of the body based on that.  The job itself is simple:
```
var body = bodies[index];
factor[index] = Unity.Burst.CompilerServices.Hint.Unlikely(occluded[index]) ?
    0 : sunFluxAtBody[index] * body.radius * body.radius * 0.5 * body.albedo;
```
This calculation is going to be re-used for every vessel, for every substep, to determine the albedo flux on that vessel during that substep.  So it actually was a lot more intuitive, to me, to convert this weird thing into a flux originating at the body.  And then re-use as much as possible the flux-to-irradiance computation.  That's why `BodyDirectIrradiances` can now be its own job covering 3 things, and `AlbedoIrradianceAtVesselJob` isn't in there only because of the sun-body-vessel position-dependent calc.

How many allocations and context switches and jobs do you believe were scheduled here?
(In most tests, it's ending up < 4/frame, most of which are 2 fixedUpdate cycles)

Following the code structure I can see is an issue.  It's one I've been struggling with generally.  Part of it is creating the jobs takes up a lot of screen space, so it's harder to see flow at a glance.  Part of it is the variable allocations at the jobs that use them -- this isn't really necessary, though it can be a micro-optimization (and would indeed be a premature one here!) if you launch jobs while you allocate and schedule for others.  It's that way now more because organizationally it was easier... tho still ugly.  Part of it is naming is hard, and as I reorganized my own work, some of the naming may not have kept up well.  (BodyDirectIrradiances ... are not the direct irradiance anymore, oops)

I agree that somewhere in the file, there needs to be a summary of what the code flow is.  If each of those job creations were turned into a 1-line method call, I think it would be a lot more clear.  (Instead, they're a multi-line method, basically...)

Reply:
I took another (... and then another) shot at restructuring the code flow and organization.